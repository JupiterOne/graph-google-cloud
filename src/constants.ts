import { PermissionErrorHandlingOptions } from './types';

/**
 * ANY_RESOURCE is used to describe any Google Cloud resource.
 * This includes all assets both ingested and not ingested with this integration
 */
export const ANY_RESOURCE = 'resource';

export const DEFAULT_FETCH_AUTHORIZATION_HANDLING_OPTIONS: PermissionErrorHandlingOptions =
  {
    throwMissingAuthPermissionError: false,
    publishMissingPermissionWarnEvent: true,
  };

export const DEFAULT_FETCH_BASIC_API_CALL_AUTHORIZATION_HANDLING_OPTIONS = {
  throwMissingAuthPermissionError: true,
  publishMissingPermissionWarnEvent: false,
};
