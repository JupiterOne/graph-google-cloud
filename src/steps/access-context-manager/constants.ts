export const STEP_ACCESS_CONTEXT_MANAGER_ACCESS_POLICIES =
  'fetch-access-context-manager-access-policies';
export const STEP_ACCESS_CONTEXT_MANAGER_ACCESS_LEVELS =
  'fetch-access-context-manager-access-levels';
export const STEP_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETERS =
  'fetch-access-context-manager-service-perimeters';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY = 'AccessPolicy';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY =
  'google_access_context_manager_access_policy';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL = 'Ruleset';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL =
  'google_access_context_manager_access_level';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER =
  'Configuration';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER =
  'google_access_context_manager_service_perimeter';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_EGRESS_POLICY =
  'ControlPolicy';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_EGRESS_POLICY =
  'google_access_context_manager_service_perimeter_egress_policy';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_INGRESS_POLICY =
  'ControlPolicy';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_INGRESS_POLICY =
  'google_access_context_manager_service_perimeter_ingress_policy';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_API_OPERATION =
  'Configuration';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_API_OPERATION =
  'google_access_context_manager_service_perimeter_api_operation';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_METHOD_SELECTOR =
  'Configuration';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_METHOD_SELECTOR =
  'google_access_context_manager_service_perimeter_method_selector';

export const RELATIONSHIP_TYPE_ACCESS_POLICY_HAS_ACCESS_LEVEL =
  'google_access_context_manager_access_policy_has_level';
export const RELATIONSHIP_TYPE_ACCESS_POLICY_HAS_SERVICE_PERIMETER =
  'google_access_context_manager_access_policy_has_service_perimeter';
export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_PROTECTS_PROJECT =
  'google_access_context_manager_service_perimeter_protects_project';
export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_LIMITS_API_SERVICES =
  'google_access_context_manager_service_perimeter_limits_api_service';

export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_HAS_EGRESS_POLICY =
  'google_access_context_manager_service_perimeter_has_egress_policy';
export const RELATIONSHIP_TYPE_EGRESS_POLICY_HAS_API_OPERATION =
  'google_access_context_manager_service_perimeter_egress_policy_has_api_operation';
export const RELATIONSHIP_TYPE_API_OPERATION_HAS_METHOD_SELECTOR =
  'google_access_context_manager_service_perimeter_api_operation_has_method_selector';
export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_HAS_INGRESS_POLICY =
  'google_access_context_manager_service_perimeter_has_ingress_policy';
export const RELATIONSHIP_TYPE_INGRESS_POLICY_HAS_API_OPERATION =
  'google_access_context_manager_service_perimeter_ingress_policy_has_api_operation';

export const IngestionSources = {
  ACCESS_CONTEXT_MANAGER_ACCESS_POLICIES: 'acm-access-policies',
  ACCESS_CONTEXT_MANAGER_ACCESS_LEVELS: 'acm-access-levels',
  ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETERS: 'acm-service-perimeters',
};

export const AccessContextManagerIngestionConfig = {
  [IngestionSources.ACCESS_CONTEXT_MANAGER_ACCESS_POLICIES]: {
    title: 'Google Cloud Access Context Manager Access Policies',
    description: 'Defines secure access boundaries in GCP.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ACCESS_CONTEXT_MANAGER_ACCESS_LEVELS]: {
    title: 'Google Cloud Access Context Manager Access Levels',
    description: 'Manages access via hierarchical levels in GCP.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETERS]: {
    title: 'Google Cloud Access Context Manager Service Perimeters',
    description: 'Secures resources within network boundaries.',
    defaultsToDisabled: false,
  },
};
