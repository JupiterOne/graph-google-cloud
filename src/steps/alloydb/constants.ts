export const STEP_ALLOYDB_POSTGRE_SQL_SERVICE = 'create-postgre-sql-service';
export const STEP_ALLOYDB_POSTGRE_SQL_CLUSTER =
  'fetch-alloydb-postgre-sql-cluster';
export const STEP_ALLOYDB_POSTGRE_SQL_INSTANCE =
  'fetch-alloydb-postgre-sql-instance';
export const STEP_ALLOYDB_POSTGRE_SQL_CONNECTION =
  'fetch-alloydb-postgre-sql-connection';
export const STEP_ALLOYDB_POSTGRE_SQL_BACKUP =
  'fetch-alloydb-postgre-sql-backup';

// Entity type
export const ENTITY_TYPE_ALLOYDB_POSTGRE_SQL_SERVICE = 'google_cloud_alloydb';
export const ENTITY_CLASS_ALLOYDB_POSTGRE_SQL_SERVICE = ['Service'];

export const ENTITY_TYPE_POSTGRE_SQL_ALLOYDB_CLUSTER =
  'google_cloud_alloydb_cluster';
export const ENTITY_CLASS_POSTGRE_SQL_CLUSTER = [
  'Database',
  'DataStore',
  'Cluster',
];

export const ENTITY_TYPE_POSTGRE_SQL_ALLOYDB_INSTANCE =
  'google_cloud_alloydb_instance';
export const ENTITY_CLASS_POSTGRE_SQL_INSTANCE = [
  'Database',
  'DataStore',
  'Host',
];

export const ENTITY_TYPE_POSTGRE_SQL_CONNECTION =
  'google_cloud_alloydb_connection';
export const ENTITY_CLASS_POSTGRE_SQL_CONNECTION = ['Network'];

export const ENTITY_TYPE_POSTGRE_SQL_BACKUP = 'google_cloud_alloydb_backup';
export const ENTITY_CLASS_POSTGRE_SQL_BACKUP = ['Backup'];

// Relationships
export const STEP_PROJECT_HAS_ALLOYDB_SERVICE_RELATIONSHIP =
  'build-project-alloydb-service-relationship';
export const STEP_PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP =
  'build-project-has-alloydb-cluster-relationship';

export const STEP_ALLOYDB_INSTANCE_USES_CLUSTER_RELATIONSHIP =
  'build-alloydb-instance-uses-cluster-relationship';

export const STEP_ALLOYDB_CLUSTER_HAS_BACKUP_RELATIONSHIP =
  'build-alloydb-cluster-has-backup-relationship';

export const STEP_ALLOYDB_INSTANCE_HAS_CONNECTION_RELATIONSHIP =
  'build-alloydb-instance-has-connection-relationship';

export const STEP_ALLOYDB_CLUSTER_USES_KMS_KEY_RELATIONSHIP =
  'build-alloydb-cluster-uses-kms-key-relationship';

export const STEP_USER_ASSIGNED_ALLOYDB_CLUSTER_RELATIONSHIP =
  'build-user-asssigned-alloydb-cluster-relatipnship';

export const RELATIONSHIP_TYPE_PROJECT_HAS_ALLOYDB_SERVICE =
  'google_cloud_project_has_alloydb';
export const RELATIONSHIP_TYPE_PROJECT_HAS_ALLOYDB_CLUSTER =
  'google_cloud_project_has_alloydb_cluster';
export const RELATIONSHIP_TYPE_ALLOYDB_INSTANCE_USES_CLUSTER =
  'google_cloud_alloydb_instance_uses_cluster';
export const RELATIONSHIP_TYPE_ALLOYDB_CLUSTER_HAS_BACKUP =
  'google_cloud_alloydb_cluster_has_backup';
export const RELATIONSHIP_TYPE_STEP_ALLOYDB_INSTANCE_HAS_CONNECTION_BACKUP =
  'google_cloud_alloydb_instance_has_connection';
export const RELATIONSHIP_TYPE_ALLOYDB_CLUSTER_USES_KMS_KEY =
  'google_cloud_alloydb_cluster_uses_kms_crypto_key';
export const RELATIONSHIP_TYPE_USER_ASSIGNED_ALLOYDB_CLUSTER =
  'google_user_assigned_cloud_alloydb_cluster';

// IngestionSources
export const IngestionSources = {
  ALLOYDB_POSTGRE_SQL_SERVICE: 'postgre-sql-service',
  ALLOYDB_POSTGRE_SQL_CLUSTER: 'alloydb-postgre-sql-cluster',
  ALLOYDB_POSTGRE_SQL_INSTANCE: 'alloydb-postgre-sql-instance',
  ALLOYDB_POSTGRE_SQL_CONNECTION: 'alloydb-postgre-sql-connection',
  ALLOYDB_POSTGRE_SQL_BACKUP: 'alloydb-postgre-sql-backup',
  PROJECT_HAS_ALLOYDB_SERVICE_RELATIONSHIP:
    'project-has-alloydb-service-relationship',
  PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP:
    'project-has-alloydb-cluster-relationship',
  ALLOYDB_INSTANCE_USES_CLUSTER_RELATIONSHIP:
    'alloydb-instance-uses-cluster-relationship',
  ALLOYDB_CLUSTER_HAS_BACKUP_RELATIONSHIP:
    'alloydb-cluster-has-backup-relationship',
  ALLOYDB_INSTANCE_HAS_CONNECTION_RELATIONSHIP:
    'alloydb-instance-has-connection-relationship',
  ALLOYDB_CLUSTER_USES_KMS_KEY_RELATIONSHIP:
    'alloydb-cluster-uses-kms-key-relationship',
  GOOGLE_USER_ASSIGNED_ALLOYDB_CLUSTER_RELATIONSHIP:
    'google-user-assigned-alloydb-cluster-relationship',
};

// IngestionSources Configs
export const AlloyDBIngestionConfig = {
  [IngestionSources.ALLOYDB_POSTGRE_SQL_SERVICE]: {
    title: 'AlloyDB for Postgre SQL service',
    description: 'AlloyDB for Postgre SQL service.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_POSTGRE_SQL_CLUSTER]: {
    title: 'AlloyDB for Postgre SQL Cluster',
    description: 'AlloyDB for Postgre SQL Cluster.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_POSTGRE_SQL_INSTANCE]: {
    title: 'AlloyDB for Postgre SQL Instance',
    description: 'AlloyDB for Postgre SQL Instance.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_POSTGRE_SQL_CONNECTION]: {
    title: 'AlloyDB for Postgre SQL Connection',
    description: 'AlloyDB for Postgre SQL Connection.',
    defaultsToDisabled: false,
  },
  [IngestionSources.PROJECT_HAS_ALLOYDB_SERVICE_RELATIONSHIP]: {
    title: 'Project AlloyDB Service Has Relationship',
    description: 'Build Project AlloyDB Service Has Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.PROJECT_HAS_ALLOYDB_CLUSTER_RELATIONSHIP]: {
    title: 'Project AlloyDB Cluster Has Relationship',
    description: 'Build Project AlloyDB Cluster Has Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_INSTANCE_USES_CLUSTER_RELATIONSHIP]: {
    title: 'AlloyDB Instance Cluster Uses Relationship',
    description: 'Build AlloyDB Instance Cluster Uses Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_CLUSTER_HAS_BACKUP_RELATIONSHIP]: {
    title: 'AlloyDB Instance Backup Has Relationship',
    description: 'Build AlloyDB Backup Cluster Has  Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_INSTANCE_HAS_CONNECTION_RELATIONSHIP]: {
    title: 'AlloyDB Instance Connection Has  Relationship',
    description: 'Build AlloyDB Instance Connection Has Relationship.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ALLOYDB_CLUSTER_USES_KMS_KEY_RELATIONSHIP]: {
    title: 'AlloyDB Cluster Kms Key Uses Relationship',
    description: 'Build AlloyDB Cluster Kms Key Has Relationship.',
    defaultsToDisabled: false,
  },
};

// IAM Permissions
export const AlloyDBPermissions = {
  STEP_ALLOYDB_POSTGRE_SQL_CLUSTER: ['alloydb.clusters.list'],
  STEP_ALLOYDB_POSTGRE_SQL_INSTANCE: ['alloydb.instances.list'],
  STEP_ALLOYDB_POSTGRE_SQL_CONNECTION: ['alloydb.instances.connect'],
  STEP_ALLOYDB_POSTGRE_SQL_BACKUP: ['alloydb.backups.get'],
};
