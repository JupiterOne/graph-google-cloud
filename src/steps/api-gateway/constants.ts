export const STEP_API_GATEWAY_APIS = 'fetch-api-gateway-apis';
export const STEP_API_GATEWAY_API_CONFIGS = 'fetch-api-gateway-api-configs';
export const STEP_API_GATEWAY_GATEWAYS = 'fetch-api-gateway-gateways';

export const ENTITY_CLASS_API_GATEWAY_API = 'Service';
export const ENTITY_TYPE_API_GATEWAY_API = 'google_api_gateway_api';

export const ENTITY_CLASS_API_GATEWAY_API_CONFIG = 'Configuration';
export const ENTITY_TYPE_API_GATEWAY_API_CONFIG =
  'google_api_gateway_api_config';

export const ENTITY_CLASS_API_GATEWAY_GATEWAY = 'Gateway';
export const ENTITY_TYPE_API_GATEWAY_GATEWAY = 'google_api_gateway_gateway';

export const RELATIONSHIP_TYPE_API_GATEWAY_API_USES_CONFIG =
  'google_api_gateway_api_uses_config';
export const RELATIONSHIP_TYPE_API_GATEWAY_API_HAS_GATEWAY =
  'google_api_gateway_api_has_gateway';
export const RELATIONSHIP_TYPE_API_GATEWAY_API_CONFIG_USES_SERVICE_ACCOUNT =
  'google_api_gateway_api_config_uses_iam_service_account';

export const IngestionSources = {
  API_GATEWAY_APIS: 'api-gateway-apis',
  API_GATEWAY_API_CONFIGS: 'api-gateway-api-configs',
  API_GATEWAY_GATEWAYS: 'api-gateway-gateways',
};

export const ApiGatewayIngestionConfig = {
  [IngestionSources.API_GATEWAY_APIS]: {
    title: 'Google Cloud API Gateway APIs',
    description: 'Endpoint management for API gateways.',
    defaultsToDisabled: false,
  },
  [IngestionSources.API_GATEWAY_API_CONFIGS]: {
    title: 'Google Cloud API Gateway API Configurations',
    description: 'Config settings for API interfaces.',
    defaultsToDisabled: false,
  },
  [IngestionSources.API_GATEWAY_GATEWAYS]: {
    title: 'Google Cloud API Gateway Gateways',
    description: 'Networking gateways for API management.',
    defaultsToDisabled: false,
  },
};

export const ApiGatewayPermissions = {
  STEP_API_GATEWAY_APIS: [
    'apigateway.apis.getIamPolicy',
    'apigateway.apis.list',
  ],
  STEP_API_GATEWAY_API_CONFIGS: [
    'apigateway.apiconfigs.list',
    'apigateway.apiconfigs.getIamPolicy',
  ],
  STEP_API_GATEWAY_GATEWAYS: [
    'apigateway.gateways.list',
    'apigateway.gateways.getIamPolicy',
  ],
};
