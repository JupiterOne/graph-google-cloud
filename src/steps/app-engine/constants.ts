export const STEP_APP_ENGINE_APPLICATION = 'fetch-app-engine-application';
export const STEP_CREATE_APP_ENGINE_BUCKET_RELATIONSHIPS =
  'build-app-engine-application-bucket-relationships';
export const STEP_APP_ENGINE_SERVICES = 'fetch-app-engine-services';
export const STEP_APP_ENGINE_VERSIONS = 'fetch-app-engine-versions';
export const STEP_APP_ENGINE_INSTANCES = 'fetch-app-engine-instances';

export const ENTITY_TYPE_APP_ENGINE_APPLICATION =
  'google_app_engine_application';
export const ENTITY_CLASS_APP_ENGINE_APPLICATION = 'Application';

export const ENTITY_TYPE_APP_ENGINE_SERVICE = 'google_app_engine_service';
export const ENTITY_CLASS_APP_ENGINE_SERVICE = 'Container';

export const ENTITY_TYPE_APP_ENGINE_VERSION = 'google_app_engine_version';
export const ENTITY_CLASS_APP_ENGINE_VERSION = 'Service';

export const ENTITY_TYPE_APP_ENGINE_INSTANCE = 'google_app_engine_instance';
export const ENTITY_CLASS_APP_ENGINE_INSTANCE = 'Host';

export const RELATIONSHIP_TYPE_APP_ENGINE_APPLICATION_HAS_SERVICE =
  'google_app_engine_application_has_service';

export const RELATIONSHIP_TYPE_APP_ENGINE_SERVICE_HAS_VERSION =
  'google_app_engine_service_has_version';

export const RELATIONSHIP_TYPE_APP_ENGINE_VERSION_HAS_INSTANCE =
  'google_app_engine_version_has_instance';

export const RELATIONSHIP_TYPE_APP_ENGINE_APPLICATION_USES_BUCKET =
  'google_app_engine_application_uses_storage_bucket';

export const RELATIONSHIP_TYPE_GOOGLE_USER_CREATED_VERSION =
  'mapping_source_created_google_user';

export const RELATIONSHIP_TYPE_SERVICE_ACCOUNT_CREATED_VERSION =
  'google_iam_service_account_created_app_engine_version';

export const IngestionSources = {
  APP_ENGINE_APPLICATION: 'app-engine-application',
  APP_ENGINE_SERVICES: 'app-engine-services',
  APP_ENGINE_VERSIONS: 'app-engine-versions',
  APP_ENGINE_INSTANCES: 'app-engine-instances',
};

export const AppEngineIngestionConfig = {
  [IngestionSources.APP_ENGINE_APPLICATION]: {
    title: 'Google Cloud App Engine Application',
    description: 'Platform for building scalable web apps.',
    defaultsToDisabled: false,
  },
  [IngestionSources.APP_ENGINE_SERVICES]: {
    title: 'Google Cloud App Engine Services',
    description: 'Modular components of App Engine apps.',
    defaultsToDisabled: false,
  },
  [IngestionSources.APP_ENGINE_VERSIONS]: {
    title: 'Google Cloud App Engine Versions',
    description: 'Versioning for App Engine application components.',
    defaultsToDisabled: false,
  },
  [IngestionSources.APP_ENGINE_INSTANCES]: {
    title: 'Google Cloud App Engine Instances',
    description: 'Running instances of App Engine services.',
    defaultsToDisabled: false,
  },
};
