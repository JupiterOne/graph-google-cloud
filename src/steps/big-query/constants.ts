export const STEP_BIG_QUERY_DATASETS = 'fetch-big-query-datasets';
export const STEP_BIG_QUERY_TABLES = 'fetch-big-query-tables';
export const STEP_BIG_QUERY_MODELS = 'fetch-big-query-models';
export const STEP_BUILD_BIG_QUERY_DATASET_KMS_RELATIONSHIPS =
  'build-big-query-dataset-kms-relationships';

export const BIG_QUERY_DATASET_ENTITY_CLASS = ['DataStore', 'Database'];
export const BIG_QUERY_DATASET_ENTITY_TYPE = 'google_bigquery_dataset';

export const BIG_QUERY_TABLE_ENTITY_CLASS = 'DataCollection';
export const BIG_QUERY_TABLE_ENTITY_TYPE = 'google_bigquery_table';

export const BIG_QUERY_MODEL_ENTITY_CLASS = 'Model';
export const BIG_QUERY_MODEL_ENTITY_TYPE = 'google_bigquery_model';

export const RELATIONSHIP_TYPE_DATASET_USES_KMS_CRYPTO_KEY =
  'google_bigquery_dataset_uses_kms_crypto_key';
export const RELATIONSHIP_TYPE_DATASET_HAS_TABLE =
  'google_bigquery_dataset_has_table';
export const RELATIONSHIP_TYPE_DATASET_HAS_MODEL =
  'google_bigquery_dataset_has_model';

export const IngestionSources = {
  BIG_QUERY_DATASETS: 'big-query-datasets',
  BIG_QUERY_MODELS: 'big-query-models',
  BIG_QUERY_TABLES: 'big-query-tables',
};

export const BigQueryIngestionConfig = {
  [IngestionSources.BIG_QUERY_DATASETS]: {
    title: 'Google Cloud BigQuery Datasets',
    description: 'Organized collections of BigQuery data.',
    defaultsToDisabled: true,
  },
  [IngestionSources.BIG_QUERY_MODELS]: {
    title: 'Google Cloud BigQuery Models',
    description: 'Machine learning models in BigQuery.',
    defaultsToDisabled: true,
  },
  [IngestionSources.BIG_QUERY_TABLES]: {
    title: 'Google Cloud BigQuery Tables',
    description: 'Structured data tables in BigQuery.',
    defaultsToDisabled: true,
  },
};
