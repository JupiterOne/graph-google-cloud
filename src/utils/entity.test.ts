import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import { storage_v1 } from 'googleapis';
import { getMockStorageBucket } from '../../test/mocks';
import { StorageEntitiesSpec } from '../steps/storage/constants';
import {
  getCloudStorageBucketKey,
  getCloudStorageBucketWebLink,
} from '../steps/storage/converters';
import { createGoogleCloudIntegrationEntity } from './entity';

function getMockStorageBucketEntityBuilderInput(
  data: storage_v1.Schema$Bucket,
  projectId: string,
) {
  return {
    entityData: {
      source: data,
      assign: {
        _class: StorageEntitiesSpec.STORAGE_BUCKET._class,
        _type: StorageEntitiesSpec.STORAGE_BUCKET._type,
        _key: getCloudStorageBucketKey(data.id as string),
        id: data.id as string,
        name: data.name,
        displayName: data.name as string,
        storageClass: data.storageClass,
        createdOn: parseTimePropertyValue(data.timeCreated),
        updatedOn: parseTimePropertyValue(data.updated),
        // Storage buckets are encrypted by default
        encrypted: true,
        // If not set, we are using the default Google Encryption key
        encryptionKeyRef: data.encryption?.defaultKmsKeyName,
        kmsKeyName: data.encryption?.defaultKmsKeyName,
        // https://cloud.google.com/storage/docs/uniform-bucket-level-access
        uniformBucketLevelAccess:
          data.iamConfiguration?.uniformBucketLevelAccess?.enabled === true,
        // 2.3 Ensure that retention policies on log buckets are configured using Bucket Lock (Scored)
        retentionPolicyEnabled: data.retentionPolicy?.isLocked,
        retentionPeriod: data.retentionPolicy?.retentionPeriod,
        retentionDate: data.retentionPolicy?.effectiveTime,
        public: false,
        // Rely on the value of the classification tag
        classification: null,
        etag: data.etag,
        webLink: getCloudStorageBucketWebLink(data, projectId),
      },
    },
  };
}

describe('#createGoogleCloudIntegrationEntity', () => {
  test('should produce entity and copy labels if labels exist on resource', () => {
    const projectId = 'j1';
    const mockStorageBucket = getMockStorageBucket({
      labels: {
        myLabel: 'test',
      },
    });

    expect(
      createGoogleCloudIntegrationEntity(
        mockStorageBucket,
        getMockStorageBucketEntityBuilderInput(mockStorageBucket, projectId),
      ),
    ).toMatchSnapshot();
  });

  test('should produce entity and not fail if labels do not exist on resource', () => {
    const projectId = 'j1';
    const mockStorageBucket = getMockStorageBucket({
      labels: undefined,
    });

    expect(
      createGoogleCloudIntegrationEntity(
        mockStorageBucket,
        getMockStorageBucketEntityBuilderInput(mockStorageBucket, projectId),
      ),
    ).toMatchSnapshot();
  });
});
