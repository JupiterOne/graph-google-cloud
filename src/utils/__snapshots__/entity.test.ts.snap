// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createGoogleCloudIntegrationEntity should produce entity and copy labels if labels exist on resource 1`] = `
{
  "_class": [
    "DataStore",
  ],
  "_key": "bucket:customer-managed-encryption-key-bucket-1234",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "encryption": {
          "defaultKmsKeyName": "projects/j1-gc-integration-dev-v2/locations/us/keyRings/j1-gc-integration-dev-v2-bucket-ring/cryptoKeys/j1-gc-integration-dev-v2-bucket-key",
        },
        "etag": "CAE=",
        "iamConfiguration": {
          "bucketPolicyOnly": {
            "enabled": false,
          },
          "uniformBucketLevelAccess": {
            "enabled": false,
          },
        },
        "id": "customer-managed-encryption-key-bucket-1234",
        "kind": "storage#bucket",
        "labels": {
          "myLabel": "test",
        },
        "location": "US",
        "locationType": "multi-region",
        "metageneration": "1",
        "name": "customer-managed-encryption-key-bucket-1234",
        "projectNumber": "1234",
        "selfLink": "https://www.googleapis.com/storage/v1/b/customer-managed-encryption-key-bucket-1234",
        "storageClass": "STANDARD",
        "timeCreated": "2020-07-28T19:06:14.033Z",
        "updated": "2020-07-28T19:06:14.033Z",
      },
    },
  ],
  "_type": "google_storage_bucket",
  "classification": null,
  "createdOn": 1595963174033,
  "displayName": "customer-managed-encryption-key-bucket-1234",
  "encrypted": true,
  "encryptionKeyRef": "projects/j1-gc-integration-dev-v2/locations/us/keyRings/j1-gc-integration-dev-v2-bucket-ring/cryptoKeys/j1-gc-integration-dev-v2-bucket-key",
  "etag": "CAE=",
  "id": "customer-managed-encryption-key-bucket-1234",
  "kmsKeyName": "projects/j1-gc-integration-dev-v2/locations/us/keyRings/j1-gc-integration-dev-v2-bucket-ring/cryptoKeys/j1-gc-integration-dev-v2-bucket-key",
  "location": "US",
  "name": "customer-managed-encryption-key-bucket-1234",
  "public": false,
  "retentionDate": undefined,
  "retentionPeriod": undefined,
  "retentionPolicyEnabled": undefined,
  "storageClass": "STANDARD",
  "tag.myLabel": "test",
  "uniformBucketLevelAccess": false,
  "updatedOn": 1595963174033,
  "webLink": "https://console.cloud.google.com/storage/browser/customer-managed-encryption-key-bucket-1234;tab=objects?forceOnBucketsSortingFiltering=false&project=j1",
}
`;

exports[`#createGoogleCloudIntegrationEntity should produce entity and not fail if labels do not exist on resource 1`] = `
{
  "_class": [
    "DataStore",
  ],
  "_key": "bucket:customer-managed-encryption-key-bucket-1234",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "encryption": {
          "defaultKmsKeyName": "projects/j1-gc-integration-dev-v2/locations/us/keyRings/j1-gc-integration-dev-v2-bucket-ring/cryptoKeys/j1-gc-integration-dev-v2-bucket-key",
        },
        "etag": "CAE=",
        "iamConfiguration": {
          "bucketPolicyOnly": {
            "enabled": false,
          },
          "uniformBucketLevelAccess": {
            "enabled": false,
          },
        },
        "id": "customer-managed-encryption-key-bucket-1234",
        "kind": "storage#bucket",
        "labels": undefined,
        "location": "US",
        "locationType": "multi-region",
        "metageneration": "1",
        "name": "customer-managed-encryption-key-bucket-1234",
        "projectNumber": "1234",
        "selfLink": "https://www.googleapis.com/storage/v1/b/customer-managed-encryption-key-bucket-1234",
        "storageClass": "STANDARD",
        "timeCreated": "2020-07-28T19:06:14.033Z",
        "updated": "2020-07-28T19:06:14.033Z",
      },
    },
  ],
  "_type": "google_storage_bucket",
  "classification": null,
  "createdOn": 1595963174033,
  "displayName": "customer-managed-encryption-key-bucket-1234",
  "encrypted": true,
  "encryptionKeyRef": "projects/j1-gc-integration-dev-v2/locations/us/keyRings/j1-gc-integration-dev-v2-bucket-ring/cryptoKeys/j1-gc-integration-dev-v2-bucket-key",
  "etag": "CAE=",
  "id": "customer-managed-encryption-key-bucket-1234",
  "kmsKeyName": "projects/j1-gc-integration-dev-v2/locations/us/keyRings/j1-gc-integration-dev-v2-bucket-ring/cryptoKeys/j1-gc-integration-dev-v2-bucket-key",
  "location": "US",
  "name": "customer-managed-encryption-key-bucket-1234",
  "public": false,
  "retentionDate": undefined,
  "retentionPeriod": undefined,
  "retentionPolicyEnabled": undefined,
  "storageClass": "STANDARD",
  "uniformBucketLevelAccess": false,
  "updatedOn": 1595963174033,
  "webLink": "https://console.cloud.google.com/storage/browser/customer-managed-encryption-key-bucket-1234;tab=objects?forceOnBucketsSortingFiltering=false&project=j1",
}
`;
